### PR TITLE
Change scheduled event generated template

### LINKED ISSUE #6640: Reordering or updating in number of Event Schedules break deployment
# This is a Bug Report

## Description

After deploying a function with `schedules`, if the order of the registered schedules change (be it because of actual ordering or because due to additions/removals a schedule with index `i` became a schedule with index `j`), further deployments can't be performed unless the stack or the stuck resources are removed.

I would rather expect that the schedules are order-agnostic and that the deployment would go through despite of the changes in the schedule

The deployment logs are below.

## Additional Data

- **_Serverless Framework Version you're using_**: 1.50.0
- **_Operating System_**: MacOS Mojave 10.14.5
- **_Stack Trace_**: 
- **_Provider Error messages_**:

```
Serverless: Packaging service...
Serverless: Excluding development dependencies...
Serverless: Excluding development dependencies...
Serverless: Excluding development dependencies...
Serverless: Uploading CloudFormation file to S3..
Serverless: Uploading artifacts...
Serverless: Uploading service two.zip file to S3 (3.08 MB)...
Serverless: Uploading service one.zip file to S3 (3.08 MB)...
Serverless: Uploading service three.zip file to S3 (3.08 MB)...
Serverless: Validating template...
Serverless: Updating Stack...
Serverless: Checking Stack update progress.
CloudFormation - UPDATE_IN_PROGRESS - AWS::CloudFormation::Stack - project-production
CloudFormation - UPDATE_IN_PROGRESS - AWS::Lambda::Function - TwoLambdaFunction
CloudFormation - UPDATE_IN_PROGRESS - AWS::Lambda::Function - OneLambdaFunction
CloudFormation - UPDATE_IN_PROGRESS - AWS::Lambda::Function - ThreeLambdaFunction
CloudFormation - UPDATE_COMPLETE - AWS::Lambda::Function - TwoLambdaFunction
CloudFormation - UPDATE_COMPLETE - AWS::Lambda::Function - ThreeLambdaFunction
CloudFormation - UPDATE_COMPLETE - AWS::Lambda::Function - OneLambdaFunction
CloudFormation - CREATE_IN_PROGRESS - AWS::Lambda::Version - TwoLambdaVersionWd3JK80dXWmbHhVFGMrJccCAx5Kodao3jGJQ5Lb8Ek
CloudFormation - CREATE_IN_PROGRESS - AWS::Lambda::Version - ThreeLambdaVersionCvqOmcZL1XCCCxef4Kk2CcJ3ok5xaFAY9WAo04XG0
CloudFormation - CREATE_IN_PROGRESS - AWS::Lambda::Version - OneLambdaVersionf1R6UE7K9gussctXIuRUWqvguba3OmqTRUTHn7FWM
CloudFormation - CREATE_IN_PROGRESS - AWS::Lambda::Version - TwoLambdaVersionWd3JK80dXWmbHhVFGMrJccCAx5Kodao3jGJQ5Lb8Ek
CloudFormation - CREATE_COMPLETE - AWS::Lambda::Version - TwoLambdaVersionWd3JK80dXWmbHhVFGMrJccCAx5Kodao3jGJQ5Lb8Ek
CloudFormation - CREATE_IN_PROGRESS - AWS::Lambda::Version - ThreeLambdaVersionCvqOmcZL1XCCCxef4Kk2CcJ3ok5xaFAY9WAo04XG0
CloudFormation - CREATE_IN_PROGRESS - AWS::Lambda::Version - OneLambdaVersionf1R6UE7K9gussctXIuRUWqvguba3OmqTRUTHn7FWM
CloudFormation - CREATE_COMPLETE - AWS::Lambda::Version - ThreeLambdaVersionCvqOmcZL1XCCCxef4Kk2CcJ3ok5xaFAY9WAo04XG0
CloudFormation - CREATE_COMPLETE - AWS::Lambda::Version - OneLambdaVersionf1R6UE7K9gussctXIuRUWqvguba3OmqTRUTHn7FWM
CloudFormation - UPDATE_IN_PROGRESS - AWS::Events::Rule - OneEventsRuleSchedule6
CloudFormation - UPDATE_IN_PROGRESS - AWS::Events::Rule - OneEventsRuleSchedule5
CloudFormation - UPDATE_IN_PROGRESS - AWS::Events::Rule - OneEventsRuleSchedule3
CloudFormation - UPDATE_FAILED - AWS::Events::Rule - OneEventsRuleSchedule6
CloudFormation - UPDATE_IN_PROGRESS - AWS::Events::Rule - OneEventsRuleSchedule7
CloudFormation - UPDATE_IN_PROGRESS - AWS::Events::Rule - OneEventsRuleSchedule4
CloudFormation - UPDATE_IN_PROGRESS - AWS::Events::Rule - OneEventsRuleSchedule1
CloudFormation - UPDATE_IN_PROGRESS - AWS::Events::Rule - OneEventsRuleSchedule2
CloudFormation - UPDATE_FAILED - AWS::Events::Rule - OneEventsRuleSchedule3
CloudFormation - UPDATE_FAILED - AWS::Events::Rule - OneEventsRuleSchedule5
CloudFormation - UPDATE_FAILED - AWS::Events::Rule - OneEventsRuleSchedule2
CloudFormation - UPDATE_FAILED - AWS::Events::Rule - OneEventsRuleSchedule1
CloudFormation - UPDATE_FAILED - AWS::Events::Rule - OneEventsRuleSchedule4
CloudFormation - UPDATE_FAILED - AWS::Events::Rule - OneEventsRuleSchedule7
CloudFormation - UPDATE_ROLLBACK_IN_PROGRESS - AWS::CloudFormation::Stack - project-production
CloudFormation - UPDATE_IN_PROGRESS - AWS::Lambda::Function - OneLambdaFunction
CloudFormation - UPDATE_IN_PROGRESS - AWS::Lambda::Function - ThreeLambdaFunction
CloudFormation - UPDATE_IN_PROGRESS - AWS::Lambda::Function - TwoLambdaFunction
CloudFormation - UPDATE_COMPLETE - AWS::Lambda::Function - TwoLambdaFunction
CloudFormation - UPDATE_COMPLETE - AWS::Lambda::Function - ThreeLambdaFunction
CloudFormation - UPDATE_COMPLETE - AWS::Lambda::Function - OneLambdaFunction
CloudFormation - UPDATE_COMPLETE - AWS::Events::Rule - OneEventsRuleSchedule5
CloudFormation - UPDATE_COMPLETE - AWS::Events::Rule - OneEventsRuleSchedule6
CloudFormation - UPDATE_COMPLETE - AWS::Events::Rule - OneEventsRuleSchedule2
CloudFormation - UPDATE_COMPLETE - AWS::Events::Rule - OneEventsRuleSchedule1
CloudFormation - UPDATE_COMPLETE - AWS::Events::Rule - OneEventsRuleSchedule3
CloudFormation - UPDATE_COMPLETE - AWS::Events::Rule - OneEventsRuleSchedule7
CloudFormation - UPDATE_COMPLETE - AWS::Events::Rule - OneEventsRuleSchedule4
CloudFormation - UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS - AWS::CloudFormation::Stack - project-production
CloudFormation - DELETE_COMPLETE - AWS::Events::Rule - OneEventsRuleSchedule6
CloudFormation - DELETE_IN_PROGRESS - AWS::Events::Rule - OneEventsRuleSchedule3
CloudFormation - DELETE_IN_PROGRESS - AWS::Events::Rule - OneEventsRuleSchedule5
CloudFormation - DELETE_IN_PROGRESS - AWS::Events::Rule - OneEventsRuleSchedule7
CloudFormation - DELETE_COMPLETE - AWS::Events::Rule - OneEventsRuleSchedule1
CloudFormation - DELETE_SKIPPED - AWS::Lambda::Version - ThreeLambdaVersionCvqOmcZL1XCCCxef4Kk2CcJ3ok5xaFAY9WAo04XG0
CloudFormation - DELETE_SKIPPED - AWS::Lambda::Version - OneLambdaVersionf1R6UE7K9gussctXIuRUWqvguba3OmqTRUTHn7FWM
CloudFormation - DELETE_SKIPPED - AWS::Lambda::Version - TwoLambdaVersionWd3JK80dXWmbHhVFGMrJccCAx5Kodao3jGJQ5Lb8Ek
CloudFormation - DELETE_COMPLETE - AWS::Events::Rule - OneEventsRuleSchedule4
CloudFormation - DELETE_COMPLETE - AWS::Events::Rule - OneEventsRuleSchedule2
CloudFormation - DELETE_COMPLETE - AWS::Events::Rule - OneEventsRuleSchedule3
CloudFormation - DELETE_COMPLETE - AWS::Events::Rule - OneEventsRuleSchedule5
CloudFormation - DELETE_COMPLETE - AWS::Events::Rule - OneEventsRuleSchedule7
CloudFormation - UPDATE_ROLLBACK_COMPLETE - AWS::CloudFormation::Stack - project-production
Serverless: Operation failed!
Serverless: View the full error output: 
```

### PR DESCRIPTION
**Pull request by gnomeria**
## What did you implement
Change the CloudWatch event schedule naming to use schedule's name in pascal case. This solved the issue on adding new event above current schedules will cause an error that stack already exists, because it appends array position in naming.

Closes #6640

## How can we verify it

```yaml
service: test-app
app: test-app

provider:
  name: aws
  runtime: python3.7
  region: ap-southeast-1
  stage: ${opt:stage, 'dev'}
  stackName: test-app-${opt:stage}
  versionFunctions: ${self:custom.lambda.versioning.${self:provider.stage}, 'false'}
  tags:
    SLS_FRAMEWORK: Yes

plugins:
  - serverless-python-requirements
functions:
  scheduler:
    name: test-app-${opt:stage}
    handler: handler.scheduler
    events: ${self:custom.${self:custom.stage}.schedules}

custom:
  stage: "${opt:stage, self:provider.stage}"
  pythonRequirements:
    dockerizePip: non-linux
  dev:
    schedules:
      - schedule:
          name: daily-schedule-1-dev
          description: "Test schedule in position 1"
          rate: cron(1 17 * * ? *)
          enabled: true
      - schedule:
          name: daily-schedule-2-dev
          description: "Test schedule in position 2"
          rate: cron(1 17 * * ? *)
          enabled: true

  prod:
    schedules:
      - schedule:
          name: daily-schedule-1-prod
          description: "Test schedule in position 1"
          rate: cron(1 17 * * ? *)
          enabled: true
      - schedule:
          name: daily-schedule-2-prod
          description: "Test schedule in position 2"
          rate: cron(1 17 * * ? *)
          enabled: true
```
From the above yml, if we add another schedule above `daily-schedule-1-dev`, it the deployment will be conflicted in the Cloudformation stack
## Todos

<details>
<summary>Useful Scripts</summary>

- `npm run test:ci` --> Run all validation checks on proposed changes
- `npm run lint:updated` --> Lint all the updated files
- `npm run lint:fix` --> Automatically fix lint problems (if possible)
- `npm run prettier-check:updated` --> Check if updated files adhere to Prettier config
- `npm run prettify:updated` --> Prettify all the updated files

</details>

- [x] Write and run all tests
- [x] Write documentation
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

### CODE REVIEW

**Comment by gnomeria** [comment]
Is the travis ci having an issue? It seems all the tests have passed but the status has not yet updated and still in pending state

File: lib/plugins/aws/lib/naming.js [hunk 1]
```diff
@@ -332,11 +335,20 @@ module.exports = {
   },
 
   // Schedule
+  normalizeScheduleName(scheduleName) {
```
Comment (medikoo) [inline lib/plugins/aws/lib/naming.js:338]: Let's not created thise utils, but use `normalizeNameToAlphaNumericOnly` instead. This is what we use in all other similar cases.

File: lib/plugins/aws/lib/naming.js [hunk 2]
```diff
@@ -332,11 +335,20 @@ module.exports = {
...
+    return this.normalizeNameToPascalCase(scheduleName);
+  },
   getScheduleId(functionName) {
     return `${functionName}Schedule`;
   },
-  getScheduleLogicalId(functionName, scheduleIndex) {
-    return `${this.getNormalizedFunctionName(functionName)}EventsRuleSchedule${scheduleIndex}`;
+  getScheduleLogicalId(functionName, scheduleIndex, scheduleName) {
+    let logicalId = `${this.getNormalizedFunctionName(functionName)}EventsRuleSchedule`;
+    if (!_.isEmpty(scheduleName)) {
```
Comment (medikoo) [inline lib/plugins/aws/lib/naming.js:346]: Let's put it simply as `if (scheduleName)`

File: lib/plugins/aws/lib/naming.js (hunk 1 above)
Comment (gnomeria) [reply lib/plugins/aws/lib/naming.js:338]: As per suggestion, using the `normalizeNameToAlphanumericOnly` would convert `your-scheduled-event-name` into `Yourscheduledeventname`. Would that be acceptable?

Comment (medikoo) [reply lib/plugins/aws/lib/naming.js:338]: Yes, we use it for other similar cases, so I believe we should not introduce exception here.

I personally also would prefer convertion to camelCase, but that should probably be done globally for all cases and possibly with major bump

**Comment by gnomeria** [comment]
Really sorry for the super late adjustments. I've made the adjustments per your suggestions. Thanks

**Comment by gnomeria** [comment]
> It looks as valuable fix, and we'll be happy to take. Still please see suggestions I've posted.
> 
> Also can you confirm it won't break for already deployed service (?)
> 
> Not that after update, resource names for named schedule resources will change. Question is wouldn't that crash CF deployment, same way as it currently happens with re-indexing (?)

Actually now that I've reviewed this, it's probably not backward compatible yet with the index naming schemes. Let me reproduce this first.

**Comment by pgrzesik** [comment]
Hello @gnomeria, thanks a lot for your contributions to this PR :bow: We'd like to wrap it up in the coming days, so I hope you don't mind me taking over your work and applying the finishing changes. Please let me know if you'd still like to continue your work here. Once again, your help is much appreciated :bow:

**Comment by mnapoli** [comment]
👋  this PR seems stalled for more than a year, should we close it for now, and reopen in case there is anything new?

Or should we do something else?

**Comment by pgrzesik** [comment]
I think we might close this one - it's on a `todo` list for me for a long time but there was always something with higher priority @mnapoli

**Comment by mnapoli** [comment]
👍  let's close, we can always reopen later

**Pull request closed without merging**